### ISSUE TITLE
Setting a custom endpoint of an SNS client results in EndpointError

### ISSUE THREAD

**Issue by darbio** [aws/aws-sdk-js-v3#4076]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Setting a custom endpoint of an SNS client results in `EndpointError: Endpoint URL must be a string, got object`.

### SDK version number

@aws-sdk/sns-client@3.194.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v16.18.0

### Reproduction Steps

```
// sns client
let config: SNSClientConfig = {};

if (process.env.IS_OFFLINE) {
    config.endpoint = "http://127.0.0.1:4002";
    config.region = "us-east-1";
}

const client = new SNSClient(config);

console.log(client.config.endpoint);
// Prints:
// [AsyncFunction (anonymous)]

console.log(await client.config.endpoint());
// Prints:
// {
//     hostname: '127.0.0.1',
//     port: 4002,
//     protocol: 'http:',
//     path: '/',
//     query: undefined
//   }

// create the queue
const { TopicArn } = await client.send(new CreateTopicCommand({
    Name: process.env.STACK_SNS_TOPIC_NAME,
    Attributes: {
        DisplayName: "Some Name",
        FifoTopic: "true"
    }
}));

// throws
// EndpointError: Endpoint URL must be a string, got object
```

### Observed Behavior

Throws an EndpointError: Endpoint URL must be a string, got object

### Expected Behavior

Client uses custom URL

**Comment by samuelabreu** [comment]
Same to me trying to use a custom endpoint on CognitoIdentityProviderClient, sticking to 3.171.0 works fine.

**Comment by only4lee** [comment]
bump. Confirm that this also happens with @aws-sdk/client-sqs@3.194.0
When I revert back to 3.171.0 as per @samuelabreu it works fine.

**Comment by eau-de-la-seine** [comment]
Hi, I have encountered the same issue with the latest `aws-sdk` version (`3.194.0`):
```
EndpointError: Endpoint URL must be a string, got object

      at getEndpointUrl (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/getEndpointUrl.js:17:11)
      at evaluateEndpointRule (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateEndpointRule.js:26:50)
      at evaluateRules (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateRules.js:11:89)
      at evaluateTreeRule (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateTreeRule.js:12:46)
      at evaluateRules (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateRules.js:20:81)
      at evaluateTreeRule (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateTreeRule.js:12:46)
      at evaluateRules (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateRules.js:20:81)
      at evaluateTreeRule (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateTreeRule.js:12:46)
      at evaluateRules (node_modules/@aws-sdk/util-endpoints/dist-cjs/utils/evaluateRules.js:20:81)
      at resolveEndpoint (node_modules/@aws-sdk/util-endpoints/dist-cjs/resolveEndpoint.js:26:48)
```

I think that the issue is global. We init the AWS client, and then there's is internal typing issue when using the methods.

Current solution: sticking to a working version like `3.193.0`

**Comment by JamesKyburz** [comment]
@eau-de-la-seine I think the issue is global too, I got the same issue with the @aws-sdk/client-eventbridge client@3.194.0, however I had to go back to version 3.190.0 for it to work with a custom endpoint.

**Comment by JamesKyburz** [comment]
3.190.0 created some typescript mismatch issues, so sticking with 3.188.0 for now.

**Comment by structurr** [comment]
Wow I can not believe I just spent two whole days thinking I'm doing something wrong with my DynamoDB config 😪

**Comment by piersatdoshii** [comment]
Hit this issue with dynamodb too, downgraded to `3.193.0` as a workaround.

**Comment by kenny-alvizuris-cko** [comment]
I hit the issue today with `@aws-sdk/client-sqs` version `3.194.0` will post once I get a stable version.

**Comment by SishaarRao** [comment]
Encountered this issue with `@aws-sdk/client-sqs` version `3.196.0`, downgraded to `3.193.0` and issue went away

**Comment by kuhe** [comment]
Thanks for reporting this issue, a fix was included in today's release of version [3.197.0](https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.197.0).
